mesonbuild/mcompile.py

```diff
@@ -0,0 +1,127 @@
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Entrypoint script for backend agnostic compile."""
+
+import os
+import pathlib
+import shutil
+import subprocess
+import sys
+import typing as T
+
+from . import mlog
+
+if T.TYPE_CHECKING:
+    import argparse
+
+
+def add_arguments(parser: 'argparse.ArgumentParser') -> None:
+    """Add compile specific arguments."""
+    parser.add_argument(
+        '-j', '--jobs',
+        action='store',
+        default=0,
+        type=int,
+        help='The number of worker jobs to run (if supported). If the value is less than 1 the build program will guess.'
+    )
+    parser.add_argument(
+        '-l', '--load-average',
+        action='store',
+        default=0,
+        type=int,
+        help='The system load average to try to maintain (if supported)'
+    )
+    parser.add_argument(
+        '--debug',
+        action='store_true',
+        help='Print debug information for the meson compile command.'
+    )
+    parser.add_argument(
+        'builddir',
+        type=pathlib.Path,
+        nargs='?',
+        default='.',
+        help='The directory containing build files to be built.'
+    )
+
+
+def _run(options: 'argparse.Namespace') -> int:
+    if options.debug:
+        mlog.log_file = sys.stderr
+        mlog.set_verbose()
+
+    bdir = options.builddir  # type: pathlib.Path
+    if not bdir.exists():
+        mlog.error('Path to builddir {} does not exist!'.format(bdir.as_posix()))
+    if not bdir.is_dir():
+        mlog.error('builddir path should be a directory.')
+
+    cmd = []  # type: T.List[str]
+    runner = None  # type T.Optional[str]
+    if (bdir / 'build.ninja').exists():
+        mlog.debug('Found build.ninja, using ninja.')
+        runner = os.environ.get('NINJA')
+        mlog.debug('Got runner {} from environment variable NINJA'.format(runner))
+        if not runner:
+            if shutil.which('ninja'):
+                runner = 'ninja'
+            elif shutil.which('samu'):
+                runner = 'samu'
+            mlog.debug('found runner {}'.format(runner))
+
+        if runner is None:
+            mlog.error('Cannot find either ninja or samu.')
+            return 1
+
+        cmd = [runner, '-C', bdir.as_posix()]
+
+        # If the value is set to < 1 then don't set anything, which let's
+        # ninja/samu decide what to do.
+        if options.jobs > 0:
+            cmd.extend(['-j', str(options.jobs)])
+        if options.load_average > 0:
+            cmd.extend(['-l', str(options.load_average)])
+
+    # TODO: with python 3.8 this could be `elif sln := bdir.glob('*.sln'):`
+    elif list(bdir.glob('*.sln')):
+        sln = list(bdir.glob('*.sln'))[0]
```

Comment:
You glob it twice

Reply:
Without the python 3.8 walrus operator (the TODO comment) there isn't a really good way not to glob twice. We could make the VS code the first if, but I suspect most people use ninja, we could do a nested if:
```python
if <build.ninja>:
    ...
else:
    <glob>
    if glob:
        ...
    else:  # xcode
        ...
```
Or we could lose the elif, and have two if statements, but that would mean even if we've already decided to use ninja we still have to glob. I felt like just adding a TODO comment and doubling the glob was the cleanest solution.

Reply:
What about:
```python
def _get_compile_cmd():
    if <build.ninja>:
        return _get_ninja_cmd()
    <glob>
    if glob:
        return _get_sln_cmd(glob)
    if ...
def _run():
  cmd = _get_compile_cmd()
  return subprocess.run(cmd)

```

Reply:
I noticed that there's a `Path.match()` function that is probably a better fit anyway, it's lazy (first match causes it to return true, so I used that instead for the elif

Reply:
Ok, that's good enough. I know nothing about VS backend, but are we sure any .sln file is good, or should we let user tell which one to use somehow?